Type mismatch in slices.SortFunc due to breaking change in golang.org/x/exp/slices

## Bug Report

### 1. Minimal reproduce step (Required)

- Clone the project and navigate to the file `charset/charset.go`.
- In the `GetSupportedCharsets()` function, the `slices.SortFunc` function is used to sort a slice of `*Charset` by the `Name` field.

### 2. What did you expect to see? (Required)

Expected the code to compile without errors and the sorting functionality using `slices.SortFunc` to work correctly with `Charset` objects.

### 3. What did you see instead? (Required)

The compilation failed with the following error message:
```
charset/charset.go:94:28: type func(i *Charset, j *Charset) int of func(i, j *Charset) int {…} does not match inferred type func(a *Charset, b *Charset) bool for func(a E, b E) bool external/com_github_go_mysql_org_go_mysql/replication/binlogsyncer.go:427
```

This error arises due to a breaking change in `golang.org/x/exp/slices`, where the signature of `slices.SortFunc` was changed from a `less function` (returning a boolean) to a `comparison function` (returning an integer). This change aligns the `exp/slices` package with the `slices` package introduced in Go 1.21.

### 4. Cause of the issue

The root cause of the issue is that the `golang.org/x/exp/slices` package was [updated](https://github.com/golang/exp/commit/302865e7556b4ae5de27248ce625d443ef4ad3ed) to match the Go 1.21 standard library `slices` package. This introduces a breaking change that affects projects using Go 1.19 (which lacks the `slices` package in the standard library) and relying on `golang.org/x/exp/slices`.

### 5. The fix (Suggested)
```
diff --git a/pkg/parser/charset/charset.go b/pkg/parser/charset/charset.go
index 2be7b9c21d..056c54ef9b 100644
--- a/pkg/parser/charset/charset.go
+++ b/pkg/parser/charset/charset.go
@@ -100,8 +100,8 @@ func GetSupportedCharsets() []*Charset {
 	}
 
 	// sort charset by name.
-	slices.SortFunc(charsets, func(i, j *Charset) int {
-		return strings.Compare(i.Name, j.Name)
+	slices.SortFunc(charsets, func(i, j *Charset) bool {
+		return strings.Compare(i.Name, j.Name) < 0
 	})
 	return charsets
 }
```
note: the [golang.org/x/exp](https://github.com/golang/exp) package was created to try experimentations and explicitly states that new versions may break existing code:

> Warning: Packages here are experimental and unreliable. Some may one day be promoted to the main repository or other subrepository, or they may be modified arbitrarily or even disappear altogether.

### 6. What is your TiDB version? (Required)
N/A

--- a/pkg/parser/charset/BUILD.bazel
+++ b/pkg/parser/charset/BUILD.bazel
@@ -20,7 +20,6 @@ go_library(
         "//pkg/parser/terror",
         "@com_github_pingcap_errors//:errors",
         "@com_github_pingcap_log//:log",
-        "@org_golang_x_exp//slices",
         "@org_golang_x_text//encoding",
         "@org_golang_x_text//encoding/charmap",
         "@org_golang_x_text//encoding/japanese",

--- a/pkg/parser/charset/charset.go
+++ b/pkg/parser/charset/charset.go
@@ -14,6 +14,7 @@
 package charset
 
 import (
+	"slices"
 	"strings"
 
 	"github.com/pingcap/errors"
@@ -21,7 +22,6 @@ import (
 	"github.com/pingcap/tidb/pkg/parser/mysql"
 	"github.com/pingcap/tidb/pkg/parser/terror"
 	"go.uber.org/zap"
-	"golang.org/x/exp/slices"
 )
 
 var (

--- a/pkg/parser/go.mod
+++ b/pkg/parser/go.mod
@@ -1,6 +1,6 @@
 module github.com/pingcap/tidb/pkg/parser
 
-go 1.19
+go 1.21
 
 require (
 	github.com/cznic/mathutil v0.0.0-20181122101859-297441e03548

--- a/pkg/parser/goyacc/BUILD.bazel
+++ b/pkg/parser/goyacc/BUILD.bazel
@@ -14,7 +14,6 @@ go_library(
         "@com_github_cznic_sortutil//:sortutil",
         "@com_github_cznic_strutil//:strutil",
         "@com_github_pingcap_errors//:errors",
-        "@org_golang_x_exp//slices",
         "@org_modernc_parser//yacc",
         "@org_modernc_y//:y",
     ],

--- a/pkg/parser/goyacc/main.go
+++ b/pkg/parser/goyacc/main.go
@@ -135,13 +135,13 @@ import (
 	"log"
 	"os"
 	"runtime"
+	"slices"
 	"sort"
 	"strings"
 
 	"github.com/cznic/mathutil"
 	"github.com/cznic/sortutil"
 	"github.com/cznic/strutil"
-	"golang.org/x/exp/slices"
 	parser "modernc.org/parser/yacc"
 	"modernc.org/y"
 )

--- a/pkg/parser/types/BUILD.bazel
+++ b/pkg/parser/types/BUILD.bazel
@@ -15,7 +15,6 @@ go_library(
         "//pkg/parser/mysql",
         "//pkg/parser/terror",
         "@com_github_cznic_mathutil//:mathutil",
-        "@org_golang_x_exp//slices",
     ],
 )
 

--- a/pkg/parser/types/field_type.go
+++ b/pkg/parser/types/field_type.go
@@ -17,6 +17,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
+	"slices"
 	"strings"
 	"unsafe"
 
@@ -24,7 +25,6 @@ import (
 	"github.com/pingcap/tidb/pkg/parser/charset"
 	"github.com/pingcap/tidb/pkg/parser/format"
 	"github.com/pingcap/tidb/pkg/parser/mysql"
-	"golang.org/x/exp/slices"
 )
 
 // UnspecifiedLength is unspecified length.

